tests/utils.py

```diff
@@ -0,0 +1,36 @@
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Tuple
+from urllib.parse import urlencode
+from uuid import UUID
+
+
+def convert_to_serializable(value: Any) -> Any:
+    """Convert values to a serializable format."""
+    if isinstance(value, datetime):
+        return value.isoformat()
+    if isinstance(value, UUID):
+        return str(value)
+    if isinstance(value, (List, Tuple)):
+        return [convert_to_serializable(v) for v in value]
+    if isinstance(value, Dict):
+        return {k: convert_to_serializable(v) for k, v in value.items()}
+    return value
+
+
+def build_query_params(
```

Comment:
Rather than just building the query parameters, why don't we actually encode the entire URL together with them? You're using an f-string above when you could just encode the entire URL. Less risk.

Reply:
As discussed in private, it won't be implemented in this PR finally, but would be great to give up the use of f-strings to verify urls in our tests later, in another PR.